libnavigation-core/src/main/java/com/mapbox/navigation/core/RoutesSetCallback.kt

```diff
@@ -0,0 +1,107 @@
+package com.mapbox.navigation.core
+
+import com.mapbox.navigation.base.route.NavigationRoute
+
+/**
+ * Interface definition for a callback that gets notified whenever routes
+ * passed to [MapboxNavigation.setNavigationRoutes] are set or produce an error and ignored.
+ */
+interface RoutesSetCallback {
+
+    /**
+     * Invoked whenever the routes passed to [MapboxNavigation.setNavigationRoutes]
+     * are successfully set.
+     *
+     * @param result [RoutesSetCallbackSuccess] object that describes the set routes
+     */
+    fun onRoutesSetResult(result: RoutesSetCallbackSuccess)
+
+    /**
+     * Invoked whenever the routes passed to [MapboxNavigation.setNavigationRoutes]
+     * produce an error and are ignored.
+     *
+     * @param result [RoutesSetCallbackError] object that describes the failure.
+     */
+    fun onRoutesSetError(result: RoutesSetCallbackError)
+}
+
+/**
+ * Represents the result of setting routes. See [RoutesSetCallback.onRoutesSetResult].
+ *
+ * @param routes List of routes that were successfully set.
+ */
+class RoutesSetCallbackSuccess internal constructor(
+    val routes: List<NavigationRoute>
+) {
```

Comment:
Good question. I'm not sure if it's a success of a failure. So here are a bunch of options:
1. We change the API to
```
fun interface RoutesSetCallback {
    fun onRoutesSet(result: RoutesSetResult)
}
```
where `RoutesSetResult` is a sealed class with 3 subclasses: Success, PrimaryRouteNotSet, NotAllAlternativesSet (the names are very raw).
2. We leave the interface as-is and add a third method `onAlternativesFilteredOut` (and maybe rename `onRoutesSetError` into `onPrimaryRouteSetError`). But that would break API in the future and for now the third method will never be invoked.
3. We change the API to
```
fun interface RoutesSetCallback {
    fun onRoutesSet(result: RoutesSetResult)
}
```
where `RoutesSetResult` is a data class that contains all info: whether the primary route was set, which alternatives were set, which alternatives were ignored, etc. Maybe there would be a status (3 possible values for now).

I personally like the first option: I find it easier to use both when you need to just check the status and get the additional info.
WDYT? @LukasPaczos

Reply:
https://github.com/mapbox/mapbox-navigation-android/pull/6040